File: src/BinaryClassificationJIT/train.cpp
#include <iostream>
#include <fstream>
#include <cstdio>
#include <iterator>
#include <limits>
#include <iomanip>
#include <chrono>
#include <signal.h>
#include <filesystem>

#include <torch/optim/adam.h>
#include <torch/csrc/jit/serialization/export.h>
#include <torch/csrc/jit/frontend/tracer.h>
#include <torch/script.h>
#include <torch/torch.h>

#include "hyperparameters.hpp"
#include "../raw_data_writer/raw_data_writer.hpp"

int main(int argc, const char* argv[]) {

    RawDataWriter writer;

    // ================== TRAINED MODEL FILE NAMES ==================

    std::string BEST_VALIDATION_MODEL = "checkpoint_v_mcc_model.pth";
    std::string BEST_TESTING_MODEL    = "checkpoint_t_mcc_model.pth";

    // ================== TRAINING DATASET ==================

    auto raw_training_dataset = torch::data::datasets::MNIST("data/MNIST/raw", torch::data::datasets::MNIST::Mode::kTrain);
    
    auto training_dataset = raw_training_dataset
        .map(torch::data::transforms::Normalize<>(0.5, 0.5))
        .map(torch::data::transforms::Stack<>());

    // RandomSampler (shuffle = true) vs SequentialSampler(shuffle = false)
    auto training_dataloader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>(
        std::move(training_dataset),
        torch::data::DataLoaderOptions()
            .batch_size(hyperparams::mini_batch_size)
            .workers(1)
    );

    // ================== VALIDATION DATASET ==================

#if defined(VALIDATION_ENABLED)

    auto raw_validation_dataset = torch::data::datasets::MNIST("data/MNIST/raw", torch::data::datasets::MNIST::Mode::kTest);
    
    auto validation_dataset = raw_validation_dataset
        .map(torch::data::transforms::Normalize<>(0.5, 0.5))
        .map(torch::data::transforms::Stack<>());

    auto validation_dataloader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>(
        std::move(validation_dataset),
        torch::data::DataLoaderOptions()
            .batch_size(hyperparams::mini_batch_size)
            .workers(1)
    );

#endif

    // ================== TEST DATASET ==================

#if defined(TESTING_ENABLED)

    auto raw_test_dataset = torch::data::datasets::MNIST("data/MNIST/raw", torch::data::datasets::MNIST::Mode::kTest);
    
    auto test_dataset = raw_test_dataset
        .map(torch::data::transforms::Normalize<>(0.5, 0.5))
        .map(torch::data::transforms::Stack<>());

    auto test_dataloader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>(
        std::move(test_dataset),
        torch::data::DataLoaderOptions()
            .batch_size(hyperparams::mini_batch_size)
            .workers(1)
    );

#endif

    // ================== LOAD CHECKPOINT IF EXIST ==================

    torch::jit::script::Module model;

    if (argc == 2) {
        try {
            model = torch::jit::load(argv[1]);
        } catch (const c10::Error& e) {
            std::cerr << "error loading the provided JIT model\n";
            return -1;
        }
    } else if (std::filesystem::exists(BEST_TESTING_MODEL) || std::filesystem::exists(BEST_VALIDATION_MODEL)) {
        
        std::cout << "Checkpoint found, loading model...\n";

        try {
            if (std::filesystem::exists(BEST_TESTING_MODEL)) {
                model = torch::jit::load(BEST_TESTING_MODEL);
            } else {
                model = torch::jit::load(BEST_VALIDATION_MODEL);
            }
        } catch (const c10::Error& e) {
            std::cerr << "Error loading the detected JIT model\n";
            return -1;
        }

        std::cout << "Checkpoint loaded successfully\n";

    } else {
        std::cout << "Jit'ed model not found.\n";
        return 0;
    }

    std::cout << '\n';

    // ================== INITIALIZE MODEL, OPTIMIZER, LOSS FUNCTION ==================
    
    std::vector<at::Tensor> parameters;
    for (const auto& params : model.parameters()) {
        parameters.push_back(params);
    }

    torch::optim::Adam optimizer(parameters, torch::optim::AdamOptions(hyperparams::learning_rate));
    torch::nn::CrossEntropyLoss loss_function;

    // ================== GET BEST AVAILABLE COMPUTING DEVICE ==================

    torch::Device device(torch::kCPU);

    if (torch::cuda::is_available()) {
        device = torch::Device(torch::kCUDA);
    } else if (torch::mps::is_available()) {
        device = torch::Device(torch::kMPS);
    }

    model.to(device);
    model.eval();

    std::cout << "Computing Device: " << device << '\n';

    // ================ SIMPLE BASIC OUTPUT TEST ==================

    std::vector<torch::jit::IValue> initial_test_input;
    initial_test_input.push_back(torch::ones({1, 1, 28, 28}));

    // batch.data | Shape: [Batch x Channel x 28 x 28]
    auto initial_test_output = model.forward(initial_test_input).toTensor();
    std::cout << "initial_test_output :\n" << initial_test_output << "\n";
    std::cout << "argmax = " << initial_test_output.argmax(1) << "\n\n";

    // ================== DISPLAY SOME DATASET INFO ==================

    int total_training_mini_batches = std::distance(training_dataloader->begin(), training_dataloader->end());
    std::cout << "training_dataset.size() = " << raw_training_dataset.size().value_or(0) <<'\n';
    std::cout << "training_loader batches = " << total_training_mini_batches <<"\n\n";

#if defined(VALIDATION_ENABLED)

    int total_validation_mini_batches = std::distance(validation_dataloader->begin(), validation_dataloader->end());
    std::cout << "validation_dataset.size() = " << raw_validation_dataset.size().value_or(0) <<'\n';
    std::cout << "validation_loader batches = " << total_validation_mini_batches <<"\n\n";

#endif

#if defined(TESTING_ENABLED)

    int total_test_mini_batches = std::distance(test_dataloader->begin(), test_dataloader->end());
    std::cout << "test_dataset.size() = " << raw_test_dataset.size().value_or(0) <<'\n';
    std::cout << "test_loader batches = " << total_test_mini_batches <<"\n\n";

#endif

    // ================== START OF TRAINING EPOCHS ==================

    auto start_time = std::chrono::high_resolution_clock::now();

#if defined(VALIDATION_ENABLED)

    float validation_best_loss = std::numeric_limits<float>::max();
    float test_best_loss       = std::numeric_limits<float>::max();

#endif

    for (int epoch_idx = 0; epoch_idx < hyperparams::training_epochs; epoch_idx++) {

        // ================== TRAINING THE MODEL ==================

#if defined(VALIDATION_ENABLED)

        float training_amassed_loss = 0.f;
        float training_correct_pred = 0.f;

        int training_overall_samples_done = 0;
        int training_current_log_samples_done = 0;
        int training_log_counts = 0;
        
        int training_mini_batch_idx = 0;
        
#endif

        for (torch::data::Example<>& training_mini_batch : *training_dataloader) {
            
            training_mini_batch.data = training_mini_batch.data.to(device);
            training_mini_batch.target = training_mini_batch.target.to(device);

            model.to(device);
            model.train();
            optimizer.zero_grad();

            std::vector<torch::jit::IValue> jit_iv_training_inputs;
            jit_iv_training_inputs.push_back(training_mini_batch.data);

            // batch.data | Shape: [Batch x Channel x 28 x 28]
            auto training_mini_batch_output = model.forward(jit_iv_training_inputs).toTensor();

            // output       | Shape: [Batch, 10]
            // batch.target | Shape: [Batch]
            // BinaryCrossEntropyLoss
            auto training_mini_batch_loss = loss_function(training_mini_batch_output, training_mini_batch.target);
            training_mini_batch_loss.backward();
            optimizer.step();

#if defined(VALIDATION_ENABLED)

            // accumulate training data every mini batches

            training_amassed_loss += training_mini_batch_loss.item<float>();
            training_correct_pred += training_mini_batch_output.argmax(1).eq(training_mini_batch.target).sum().item<int>();
            training_overall_samples_done += training_mini_batch.data.size(0);
            training_current_log_samples_done += training_mini_batch.data.size(0);
            training_log_counts++;

            // validate the network after a certain mini batch number

            if (!((training_mini_batch_idx + 1) % hyperparams::mini_batch_log_interval) || training_mini_batch_idx == 0) {

                // calculate running averages of training data

                auto training_running_loss = training_amassed_loss / static_cast<float>(training_log_counts);
                auto training_running_accuracy = training_correct_pred / training_current_log_samples_done;

                writer.add_scalar("Training Loss", training_running_loss, epoch_idx * total_training_mini_batches + training_mini_batch_idx);
                writer.add_scalar("Training Accuracy", training_running_accuracy, epoch_idx * total_training_mini_batches + training_mini_batch_idx);

                std::printf(
                    "Training: Batch [%d/%d] | Epoch : %d/%d | Accuracy (%d/%d) : %.2f%% | Loss: %.7f | Dataset: [%5d/%5d]\n",
                    training_mini_batch_idx + 1,
                    total_training_mini_batches,
                    epoch_idx + 1,
                    hyperparams::training_epochs,
                    static_cast<int>(training_correct_pred),
                    training_current_log_samples_done,
                    training_running_accuracy * 100,
                    training_running_loss,
                    training_overall_samples_done,
                    static_cast<int>(raw_training_dataset.size().value_or(0))
                );

                // ================== VALIDATING THE MODEL ==================

                float validation_running_loss = 0.f;
                float validation_correct_pred = 0.f;
                float validation_samples_done = 0.f;

                int validation_mini_batch_idx = 0;

                for (torch::data::Example<>& validation_batch : *validation_dataloader) {
                    
                    model.eval();
                    
                    validation_batch.data = validation_batch.data.to(device);
                    validation_batch.target = validation_batch.target.to(device);

                    std::vector<torch::jit::IValue> jit_iv_validation_inputs;
                    jit_iv_validation_inputs.push_back(validation_batch.data);

                    auto validation_batch_output = model.forward(jit_iv_validation_inputs).toTensor();
                    auto validation_batch_loss = loss_function(validation_batch_output, validation_batch.target);
                
                    validation_running_loss += validation_batch_loss.item<float>();
                    validation_correct_pred += validation_batch_output.argmax(1).eq(validation_batch.target).sum().item<int>();
                    validation_samples_done += validation_batch.data.size(0);
                
                    validation_mini_batch_idx++;
                }

                auto validation_ave_loss = validation_running_loss / static_cast<float>(validation_mini_batch_idx);
                auto validation_accuracy = validation_correct_pred / validation_samples_done;
            
                writer.add_scalar("Validation Loss", validation_ave_loss, epoch_idx * total_training_mini_batches + training_mini_batch_idx);
                writer.add_scalar("Validation Accuracy", validation_accuracy, epoch_idx * total_training_mini_batches + training_mini_batch_idx);

                std::printf(
                    "Validation: Batch [%d/%d] | Epoch : %d/%d | Accuracy: %.2f%% | Loss: %.7f | Dataset: [%5d/%5d]\n\n",
                    validation_mini_batch_idx,
                    total_validation_mini_batches,
                    epoch_idx + 1,
                    hyperparams::training_epochs,
                    validation_accuracy * 100,
                    validation_ave_loss,
                    static_cast<int>(validation_samples_done),
                    static_cast<int>(raw_validation_dataset.size().value_or(0))
                );

                if (validation_ave_loss < validation_best_loss) {
                    validation_best_loss = validation_ave_loss;
                    
                    try {
                        model.to(torch::kCPU);
                        model.eval();

                        std::ofstream save_v_model(
                            BEST_VALIDATION_MODEL,
                            std::ios::trunc | std::ios::out | std::ios::binary
                        );

                        torch::jit::ExportModule(model, BEST_VALIDATION_MODEL);

                    } catch (const c10::Error& e) {
                        std::cerr << "ERROR saving best VALIDATION JIT model\n\n";
                        std::cerr << "err.msg : " << e.msg() << "\n";
                        return -1;
                    }

                    std::cout << "new best VALIDATION jit model saved!\n\n";
                }

                // compare training and validation losses
                writer.add_scalars("Training vs Validation Loss", {
                    {"Training",   training_running_loss},
                    {"Validation", validation_ave_loss}
                }, epoch_idx * total_training_mini_batches + training_mini_batch_idx);

                // compare training and validation accuracy
                writer.add_scalars("Training vs Validation Accuracy", {
                    {"Training",   training_running_accuracy},
                    {"Validation", validation_accuracy}
                }, epoch_idx * total_training_mini_batches + training_mini_batch_idx);

                // zero out accumulated training data to get the next proper running averages

                training_amassed_loss = 0.f;
                training_correct_pred = 0.f;

                training_log_counts = 0;
                training_current_log_samples_done = 0.f;

                // ================== VALIDATION : END ==================
            }
#endif

#if defined(VALIDATION_ENABLED)
            training_mini_batch_idx++;
#endif
        }

        // ================== TRAINING : END ==================

        // ================== TESTING THE MODEL ==================

#if defined(TESTING_ENABLED)

        float test_amassed_loss = 0.f;
        float test_correct_pred = 0.f;
        float test_samples_done = 0.f;

        int test_mini_batch_idx = 0;

        for (torch::data::Example<>& test_batch : *test_dataloader) {
            
            model.eval();
            
            test_batch.data = test_batch.data.to(device);
            test_batch.target = test_batch.target.to(device);

            std::vector<torch::jit::IValue> jit_iv_test_inputs;
            jit_iv_test_inputs.push_back(test_batch.data);

            auto test_batch_output = model.forward(jit_iv_test_inputs).toTensor();
            auto test_batch_loss = loss_function(test_batch_output, test_batch.target);
        
            test_amassed_loss += test_batch_loss.item<float>();
            test_correct_pred += test_batch_output.argmax(1).eq(test_batch.target).sum().item<int>();
            test_samples_done += test_batch.data.size(0);

            test_mini_batch_idx++;
        }

        auto test_ave_loss = test_amassed_loss / static_cast<float>(test_mini_batch_idx);
        auto test_accuracy = test_correct_pred / test_samples_done;
    
        writer.add_scalar("Testing Loss", test_ave_loss, epoch_idx * total_training_mini_batches + training_mini_batch_idx);
        writer.add_scalar("Testing Accuracy", test_accuracy, epoch_idx * total_training_mini_batches + training_mini_batch_idx);

        std::printf(
            "Testing: Batch [%d/%d] | Epoch : %d/%d | Accuracy: %.2f%% | Loss: %.7f | Dataset: [%5d/%5d]\n\n",
            test_mini_batch_idx,
            total_test_mini_batches,
            epoch_idx + 1,
            hyperparams::training_epochs,
            test_accuracy * 100,
            test_ave_loss,
            static_cast<int>(test_samples_done),
            static_cast<int>(raw_test_dataset.size().value_or(0))
        );

        if (test_ave_loss < test_best_loss) {
            test_best_loss = test_ave_loss;

            model.to(torch::kCPU);
            model.eval();

            try {
                std::ofstream save_v_model(
                    BEST_TESTING_MODEL,
                    std::ios::trunc | std::ios::out | std::ios::binary
                );

                torch::jit::ExportModule(model, BEST_TESTING_MODEL);

            } catch (const c10::Error& e) {
                std::cerr << "ERROR saving best TESTING JIT model\n\n";
                std::cerr << "err.msg : " << e.msg() << "\n";
                std::cerr << "err.stack : " << e.backtrace() << "\n";
                return -1;
            }

            std::cout << "new best TESTING jit model saved!\n\n";
        }
        
#endif

        // ================== TESTING : END ==================

        std::cout << '\n';
    }

    // ================== END OF TRAINING EPOCHS ==================

    auto end_time = std::chrono::high_resolution_clock::now();
    auto total_time = std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time);

    size_t seconds = total_time.count();
    float minutes = static_cast<float>(total_time.count()) / 60.f;
    float hours = minutes / 60.f;

    std::cout << "total runtime time : " << seconds << " sec(s), " << minutes << " min(s), " << hours << " hr(s)\n";

    return 0;
}

File: src/BinaryClassificationJIT/train.py
import datetime
import os
import sys
import time
import signal
import torch

from torch.utils.data import DataLoader
from torch.utils.tensorboard.writer import SummaryWriter
from torch.backends import mps

from torchvision import datasets, transforms

# define hyperparameters

LEARNING_RATE = 0.001
MINI_BATCH_SIZE = 500
MINI_BATCH_LOG_INTERVAL = 10
TRAINING_EPOCHS = 2
# MOMENTUM = None

# load training and test data

INPUT_DIM = 28 # 28x28 image

TRAIN_DATASET = datasets.MNIST(
    root="data",
    train=True,
    download=True,
    transform=transforms.Compose([
        transforms.Resize((INPUT_DIM, INPUT_DIM)),
        transforms.ToTensor()
    ])
)
TRAIN_DATALOADER = DataLoader(TRAIN_DATASET, batch_size=MINI_BATCH_SIZE, shuffle=False)

VALIDATION_DATASET = datasets.MNIST(
    root="data",
    train=False,
    download=True,
    transform=transforms.Compose([
        transforms.Resize((INPUT_DIM, INPUT_DIM)),
        transforms.ToTensor()
    ])
)
VALIDATION_DATALOADER = DataLoader(VALIDATION_DATASET, batch_size=MINI_BATCH_SIZE, shuffle=False)

TEST_DATASET = None
TEST_DATALOADER = None

# Load saved model if exist for when resuming training

BEST_VALIDATION_MODEL_SAVE_PATH = os.path.join('checkpoint_v_mcc_model.pth')
BEST_TESTING_MODEL_SAVE_PATH    = os.path.join('checkpoint_t_mcc_model.pth')

# just delete or rename the model if you want to start training again
if os.path.isfile(BEST_TESTING_MODEL_SAVE_PATH):
    print('Training Checkpoint Found: Resuming Training')
    MODEL = torch.jit.load(BEST_TESTING_MODEL_SAVE_PATH)
elif os.path.isfile(BEST_VALIDATION_MODEL_SAVE_PATH):
    print('Validation Checkpoint Found: Resuming Training')
    MODEL = torch.jit.load(BEST_VALIDATION_MODEL_SAVE_PATH)
else:
    print('Error: jit\'ed model not found!')
    exit(1)

# get best available compute device and load the model to it

DEVICE = (
         "cuda" if torch.cuda.is_available()
    else "mps"  if mps.is_available()
    else "cpu"
)

MODEL.to(DEVICE)
MODEL.eval()
print(f'Device Loaded To : {DEVICE}')

# Simple Input Test

initial_test_output = MODEL(torch.ones(1, 1, 28, 28))
print('initial_test_output :\n', initial_test_output, '\n')
print('initial_test_output :\n', initial_test_output.argmax(1), '\n\n')

# Initialize optimizer

OPTIMIZER = torch.optim.Adam(MODEL.parameters(), lr=LEARNING_RATE)

# initialize loss/cost function

# Common loss functions include nn.MSELoss (Mean Square Error) for
# regression tasks, and nn.NLLLoss (Negative Log Likelihood) for
# classification. nn.CrossEntropyLoss combines nn.LogSoftmax and nn.NLLLoss.

LOSS_FUNCTION = torch.nn.CrossEntropyLoss()

# register eventhandler to save model on exit


# defined test loop

def test_loop(dataset, dataloader):
    
    # Set the MODEL to evaluation mode - important for batch normalization and dropout layers
    
    MODEL.eval()

    total_dataset_size = len(dataset)
    total_mini_batches = len(dataloader)
    total_samples_done = 0
    average_loss, correct = 0, 0

    # Evaluating the MODEL with torch.no_grad() ensures that no gradients
    # are computed during test mode also serves to reduce unnecessary gradient
    # computations and memory usage for tensors with requires_grad=True

    with torch.no_grad():
        for batch_inputs, batch_labels in dataloader:

            batch_inputs.to(DEVICE)
            batch_labels.to(DEVICE)

            batch_output = MODEL(batch_inputs)

            average_loss += LOSS_FUNCTION(batch_output, batch_labels).item()
            correct += (batch_output.argmax(1) == batch_labels).type(torch.float).sum().item()
            total_samples_done += len(batch_labels)

    # calculate average loss and the accuracy of model during test

    average_loss /= total_mini_batches
    accuracy = correct / total_dataset_size


    print(f"test_loop  | Accuracy({int(correct)}/{int(total_samples_done)}): {100 * accuracy:>0.2f}%, Avg. Loss: {average_loss:>8f} \n")

    return accuracy, average_loss

# main loop

if __name__ == "__main__":
    LAST_EPOCH = 0

    print(f'Hyperparameters: \nLearning Rate = {LEARNING_RATE}\nMini-Batch Size = {MINI_BATCH_SIZE}\nEpochs = {LAST_EPOCH + 1}/{TRAINING_EPOCHS}')

    timestamp = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
    writer = SummaryWriter('runs/training_{}'.format(timestamp))

    # # write model graph in pytorch
    # writer.add_graph(MODEL, torch.rand(1, 1, INPUT_DIM, INPUT_DIM))

    best_validation_loss = 1_000_000.
    best_test_loss = 1_000_000.0

    # dataset info

    total_training_mini_batches = len(TRAIN_DATALOADER)
    total_training_data = len(TRAIN_DATASET)

    # per epoch training

    start_time = time.time()
    epoch_idx = 0
    
    while epoch_idx < TRAINING_EPOCHS:

        # Make sure gradient tracking is on, and do a pass over the data
        MODEL.train(True)

        # =========== TRAINING LOOP STARTS HERE =============
        
        training_amassed_loss = 0.
        training_correct_pred = 0.

        training_overall_samples_done = 0.
        training_current_log_samples_done = 0.
        training_log_counts = 0.

        for training_mini_batch_idx, training_mini_batch in enumerate(TRAIN_DATALOADER):

            # load the inputs and labels from the dataloader

            batch_inputs = training_mini_batch[0].to(DEVICE)
            batch_labels = training_mini_batch[1].to(DEVICE)

            # Set the MODEL to training mode - important for batch normalization and dropout layers

            MODEL.to(DEVICE)
            MODEL.train()

            # zero your gradients to prevent gradient accumulation

            OPTIMIZER.zero_grad()

            # feed the inputs to the network

            batch_output: torch.Tensor = MODEL(batch_inputs)

            # calculate the loss of the network

            loss = LOSS_FUNCTION(batch_output, batch_labels)

            # calculate the gradients of the network and perform backpropagation

            loss.backward()
            OPTIMIZER.step()

            # get the loss scalar value and calculate the number of processed data from the dataset

            loss, samples_processed = loss.item(), training_mini_batch_idx * MINI_BATCH_SIZE + len(batch_inputs)

            # accumulate loss in each mini-batches

            training_amassed_loss += loss
            training_correct_pred += (batch_output.argmax(1) == batch_labels).type(torch.float).sum().item()
            training_overall_samples_done += len(batch_inputs)
            training_current_log_samples_done += len(batch_inputs)
            training_log_counts += 1

            # log loss every `minibatch_log_interval`
            if (not ((training_mini_batch_idx + 1) % MINI_BATCH_LOG_INTERVAL)) or training_mini_batch_idx == 0:

                # calculate the training accuracy and loss for each minibatch log interval

                training_running_loss = training_amassed_loss / training_log_counts
                training_running_accuracy = training_correct_pred / training_current_log_samples_done

                print(f"train_loop | Epoch : {epoch_idx + 1}/{TRAINING_EPOCHS} | Accuracy({int(training_correct_pred)}/{int(training_current_log_samples_done)}): {training_running_accuracy * 100:>0.2f}% | Loss: {loss:>7f} | [{samples_processed:>5d}/{total_training_data:>5d}]")

                # Write Training Reports to Tensorboard

                writer.add_scalar(
                    f'Accuracy every {MINI_BATCH_LOG_INTERVAL} mini-batches: Training',
                    training_running_accuracy,
                    epoch_idx * total_training_mini_batches + training_mini_batch_idx
                )

                writer.add_scalar(
                    f'Loss every {MINI_BATCH_LOG_INTERVAL} mini-batches: Training',
                    training_running_loss,
                    epoch_idx * total_training_mini_batches + training_mini_batch_idx
                )

                # if there is a test dataloader perform validation runs and write reports to tensorboard

                if VALIDATION_DATASET != None and VALIDATION_DATALOADER != None:

                    #========== run validation every after training ==========#

                    validation_accuracy, validation_loss = test_loop(VALIDATION_DATASET, VALIDATION_DATALOADER)
                    
                    # Set the MODEL back to training mode after tests
                    
                    MODEL.train()

                    # write specific reports to tensorboard

                    writer.add_scalar(
                        f'Accuracy every {MINI_BATCH_LOG_INTERVAL} mini-batches: Validation',
                        validation_accuracy,
                        epoch_idx * total_training_mini_batches + training_mini_batch_idx
                    )

                    writer.add_scalars(
                        f'Accuracy every {MINI_BATCH_LOG_INTERVAL} mini-batches: Training vs Validation',
                        {
                            'Training': training_running_accuracy,
                            'Validation': validation_accuracy
                        },
                        epoch_idx * total_training_mini_batches + training_mini_batch_idx
                    )

                    writer.add_scalars(
                        f'Avg. Loss per {MINI_BATCH_LOG_INTERVAL} mini-batches: Training vs Validation',
                        {
                            'Training': training_running_loss,
                            'Validation': validation_loss
                        },
                        epoch_idx * total_training_mini_batches + training_mini_batch_idx
                    )

                    # Track best performance, and save the model's state
                    if validation_loss < best_validation_loss:

                        best_validation_loss = validation_loss

                        try:
                            MODEL.eval()
                            MODEL.to("cpu")
                            torch.jit.save(MODEL, BEST_VALIDATION_MODEL_SAVE_PATH)
                            print('New Best Validation Result Model Saved')
                        except Exception as e:
                            print(f'Error Occured: {e}')

                # zero out accumulated training data to get the next proper running averages

                training_amassed_loss = 0.0
                training_correct_pred = 0.0

                training_log_counts = 0
                training_current_log_samples_done = 0.0

        #========== TRAINING LOOP ENDS HERE ==========#

        #========== RUN TEST ON TEST DATASET EVERY EPOCH ==========#

        if TEST_DATASET != None and TEST_DATALOADER != None:
            test_accuracy, test_loss = test_loop(TEST_DATASET, TEST_DATALOADER)

            print(f'\nEpoch Test Accuracy : {test_accuracy}')
            print(f'Epoch Test Loss       : {test_loss}')

            writer.add_scalar(
                f'Test Accuracy Every Epoch',
                test_accuracy,
                epoch_idx
            )

            writer.add_scalar(
                f'Test Loss Every Epoch',
                test_loss,
                epoch_idx
            )

            if test_loss < best_test_loss:
                best_test_loss = test_loss

                try:
                    MODEL.eval()
                    MODEL.to("cpu")
                    torch.jit.save(MODEL, BEST_TESTING_MODEL_SAVE_PATH)
                    print('New Best Test Result Model Saved')
                except Exception as e:
                    print(f'Error Occured: {e}')

        writer.flush()
        epoch_idx += 1
        LAST_EPOCH = epoch_idx

    #========== END OF TRAINING EPOCH LOOP ==========#

    writer.close()
    end_time = time.time()
    print("--- %s seconds ---" % (end_time - start_time))
